oak_functions/loader/src/server.rs

```diff
@@ -486,22 +443,26 @@ fn check_export_function_signature(
     }
 }
 
-/// Runs the given function and applies the given policy to the execution and result of the
-/// function.
+/// Runs the given function and applies the given security policy to the execution of the function
+/// and the response returned from it. Serializes and returns the response as a binary
+/// protobuf-encoded byte array of a constant size.
 ///
-/// If the policy is `None`, this function just runs the given function and returns its result.
-/// If a policy is present, the given function is run, but its generated response is only returned
-/// if allowed by the policy. If the execution of the function takes longer than allowed by the
-/// policy, or the resulting response is larger than allowed by the policy, an error response
-/// generated by the policy is returned instead of the actual response.
-async fn apply_policy<F, S>(
-    policy: Policy,
-    function: F,
-) -> anyhow::Result<http::response::Response<Body>>
+/// If the execution of the `function` takes longer than allowed by the given security policy,
+/// an error response with status `PolicyTimeViolation` is returned. If the size of the `body` in
+/// the response returned by the `function` is larger than allowed by the security policy, the
+/// response is discarded and a response with status `PolicySizeViolation` is returned instead.
+/// In all cases, to keep the total size of the returned byte array constant, the `body` of the
+/// response may be padded by a number of trailing 0s before encoding the response as a binary
+/// protobuf message. In this case, the `length` in the response will contain the effective length
+/// of the `body`. For a valid security policy, this response is guaranteed to comply with the
+/// policy's size restriction. If the policy is not valid, the function returns early with an error.
+pub async fn apply_policy<F, S>(policy: Policy, function: F) -> anyhow::Result<Response>
 where
     F: std::marker::Send + 'static + FnOnce() -> S,
     S: std::future::Future<Output = anyhow::Result<Response>> + std::marker::Send,
 {
+    policy.validate()?;
```

Comment:
Another option is to have separate `Policy` and `ValidatedPolicy` types; then make `ValidatedPolicy` impl `TryFrom<Policy>` (and this should be the only way of constructing an instance, of course).